[IMP] website: introduce `s_carousel_intro` snippet

Brieuc-brd opened the pull request:
### Introduce numbered carousel indicators
This PR adds an option for carousels and `s_image_gallery` to display numbers as indicators.

### Introduce `s_carousel_intro` snippet
This PR adds the new `s_carousel_intro` snippet.
It also moves carousels options into an xml template. This was made to be able to customize these options when a carousel needs to change already existing options.
For example, this new `s_carousel_intro` snippet changes the values and label of the "Style" option.

task-4077628
Part of task-4077427

---
Requires:
- https://github.com/odoo/design-themes/pull/893
---


<img width="1487" alt="Capture d’écran 2024-09-11 à 15 12 40" src="https://github.com/user-attachments/assets/3dd63349-a79c-44af-9531-47ea4d3af92e">



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/178589.png)](https://mergebot.odoo.com/odoo/odoo/pull/178589)

stefanorigano reviewed (changes requested):
Hello @Brieuc-brd , thanks so much for handling this, good job 👍 
Some design-related remarks for you + one remark about the options.
I didn't check the CSS details since it will be affected my proposed changes anyway.

Also, since we're here, I noticed a issue with the "Invert Colors" option that behaves inconsistently **with solid buttons** (it's already like that in master).

| invert OFF | invert ON |
|--------|--------|
| target color is white ✅ ![image](https://github.com/user-attachments/assets/c5e06dee-8eb3-4d84-8c5e-ff4121ef132d) | target color is black ✅ ![image](https://github.com/user-attachments/assets/2fb91a20-998a-4a31-8d04-8a1381e89595) |
|  target color should be white, but buttons are predominately black 🤔![image](https://github.com/user-attachments/assets/cf17920e-0219-4e4d-9ef1-e4193f8b71c7) |  target color should be black but buttons are predominately white 🤔![image](https://github.com/user-attachments/assets/4487c56b-8c07-4dc6-9f60-0fb500e5e64a)  white |

The main purpose of "Invert Colors" is to make controllers visible depending by the images. The current setup leads to a scenario when, for example, the button's arrow is visible, while the button itself blends with the image underneath .

Thanks!

Brieuc-brd commented on addons/website/views/snippets/s_carousel_intro.xml:
```diff
@@ -0,0 +1,118 @@
+            <!-- Content -->
+            <div class="carousel-inner">
+                <!-- #01 -->
+                <div class="carousel-item active o_cc o_cc5 oe_img_bg o_bg_img_center px-0 pt152 pb152" style="background-image: url('/web/image/website.s_carousel_intro_default_image_1');" data-name="Slide">
+                    <div class="o_we_bg_filter bg-black-50"/>
```
This was made on purpose to handle industry images. Since we don't know which kind of images the Web Builder will define if the user add this snippet, this will ensure the text will always be visible by default.
Or maybe using a white filter ? Mhh what do you think ? 🤔

stefanorigano replied:
> This was made on purpose to handle industry images.

Right. We should then change the demo images I'm afraid... let's use others that suit better an overlay 💪

Brieuc-brd replied:
Ok, I updated the images 👍

Brieuc-brd commented on addons/website/views/snippets/s_carousel_intro.xml:
```diff
@@ -0,0 +1,118 @@
+        <attribute name="data-exclude"/>
+    </xpath>
+    <xpath expr="//we-select" position="replace">
+        <we-select string="Controllers" data-apply-to=".s_carousel_intro_controllers">
```
It is indeed much simpler 👍 Apart from the fact that we lose the possibility of having both controllers when we want to center one of them, there is still an issue with your suggestion:
- 1. hide the arrows or indicators
- 2. select “Centered"
- 3. redisplay the controller previously hidden
=> The Controllers feature will still use the “Centered” option, which should not be available, and will break the layout : https://www.awesomescreenshot.com/video/31175114?key=6c44273632d0ebf48bda83894ca3bdb5

I'm looking for a solution 👀

stefanorigano replied:
Sure. You'll have to use JS and automatically clean the DOM from unwanted classes.
Also, you should use `_computeWidgetState` to set the "_Default_" design active as soon as one of the controllers it's visible again  ;)

You may find interesting examples is `s_card/options.js` 😉

Brieuc-brd commented:
Thanks for your review @stefanorigano 👍 
I've updated the branch and simplified the "Controllers" options as you suggested 💪 

> Also, since we're here, I noticed a issue with the "Invert Colors" option that behaves inconsistently with solid buttons (it's already like that in master).

I took the opportunity to move all changes related to general improvements for carousels, including the "invert colors" option review, in another PR (https://github.com/odoo/odoo/pull/179389) and I've temporarily cherry picked the commits in this PR.

Could you have a look at them ?
Thanks in advance !

Brieuc-brd commented:
Hey @stefanorigano 👋 
I reviewed this snippet, could you have a look at it ?
Thanks in advance !

SergeBayet commented on addons/website/static/src/snippets/s_carousel_intro/options.js:
```diff
@@ -0,0 +1,28 @@
+/** @odoo-module **/
```
This is not needed anymore since this [commit](https://github.com/odoo-dev/odoo/commit/5f2c505836002ac7851c29c28d612f721d467bc4) :wink:

SergeBayet commented on addons/website/static/src/snippets/s_carousel_intro/options.js:
```diff
@@ -0,0 +1,28 @@
+            const controllersEl = this.$target[0];
+            const carouselEl = controllersEl.parentNode.parentNode;
+            const indicatorsEl = controllersEl.querySelector('.carousel-indicators');
+            if(
```
same here

SergeBayet reviewed (changes requested):
Good job :+1: Some remarks, linting mostly for the js part. Review of the design-PR ongoing.

Brieuc-brd commented:
Thanks for your review @SergeBayet 👍 
I've updated the branch, here's the diff : 
https://github.com/odoo/odoo/compare/b1c0c8e7e6dac592f823f06f5b062ea4ea459af8..f39b5f47329198cf26cc505bf0dcde9ad413e177

As discussed, I also updated the DT branch ;)

Can you take a look ?
Thanks in advance !

SergeBayet reviewed (changes requested):
Thanks for the changes made. Just a last little suggestion. See below.

Brieuc-brd commented on addons/website/static/src/snippets/s_carousel_intro/options.js:
```diff
@@ -0,0 +1,26 @@
+        // arrows and indicators are displayed.
+        if (methodName === "selectClass" && params.name === "carousel_controllers_centered_opt") {
+            const controllersEl = this.$target[0];
+            const carouselEl = controllersEl.parentNode.parentNode;
```
Thanks ! I updated the code 👍

Brieuc-brd commented:
Thanks @SergeBayet ! 👍 
I updated the branches and replied to your comments, can you take a look ?
Here are the diffs : 
- COM: https://github.com/odoo/odoo/compare/a2af3fe3d9b24fe1964cf912a03da53f3577b5ef..092012ea9eab96207f695cbeb3546037eb5f478a
- DT: https://github.com/odoo/design-themes/compare/f79d81dd05b45ae84f670a8375b4ad73dd80a384..7bcf20e74f3f6fabe5e05575aef056674770dd9f

SergeBayet reviewed (approved):
LGTM

SergeBayet commented:
Commit(s) moved into
- https://github.com/odoo/odoo/pull/180736
